_posts/2020-09-06-how-the-system-tests-works.md

```diff
@@ -0,0 +1,607 @@
+            .endInlineLogging()
+        .endKafka()
+        ...
+        ... (more configs...)
+        ... 
+    .endSpec()
+    .done()
+```
+
+In this case you can imagine that it is really easy to create multiple different instances of the Kafka instead of having separate objects.
+
+## Resources
+
+Resources in our way are abstraction to the `.yaml` definitions of custom resources which we provide such as Kafka, KafkaTopic, KafkaUser, KafkaConnect, KafkaMirrorMaker and so on.
+These custom resources are encapsulated into java objects and some of you can remind the [ORM(Object Relation Mapping)](https://www.tutorialspoint.com/hibernate/orm_overview.htm) from databases. In the following Figure 1.1 we can see YAML representation of KafkaTopic and on the Figure 1.2 related JAVA object.
+
+Figure 2.1 YAML representation of KafkaTopic
+```
+apiVersion: kafka.strimzi.io/v1beta1
+kind: KafkaTopic
+metadata:
+  name: my-topic
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  partitions: 1
+  replicas: 1
+  config:
+    retention.ms: 7200000
+    segment.bytes: 1073741824
+```
+Figure 2.2 Loading the YAML file to JAVA representation object.
+```
+KafkaTopic kafkaTopic = getKafkaTopicFromYaml(PATH_TO_KAFKA_TOPIC_CONFIG); 
+// additional changes...
+kafkaTopic = new KafkaTopicBuilder(kafkaTopic)
+    .withNewMetadata()
+        .withName(topicName)
+        .withNamespace(ResourceManager.kubeClient().getNamespace())
+        .addToLabels(Labels.STRIMZI_CLUSTER_LABEL, clusterName)
+    .endMetadata()
+    .editSpec()
+        .withPartitions(partitions)
+        .withReplicas(replicas)
+        .addToConfig("min.insync.replicas", minIsr)
+    .endSpec();
+```
+
+Everything is build with three stacks. These three stacks are managed by `ResourceManager`, which is taking care of switching. First one is `method stack` where his responsibility is to hold all
+resources, which are invoked in the test cases. Second one is the `class stack` where his responsibility is to hold all resources,
+which are invoked in the @BeforeAll notation. Switching between these two stacks are doing the third stack called pointer stack. The pointer stack points to class stack where it is in the @BeforeAll notation otherwise we are working with the method stack. The logic inside these stacks are that if you create some resource
+for instance KafkaTopic. It will be pushed inside the stack and the deletion of topic will be planned. In the teardown phase method stack will pop all resources after test case is over and same logic applied with class stack, which will pop all resources when test suite is over. What is worth mentioning is that if you specify `SKIP_TEARDOWN` environment variable to `TRUE` it will skip the teardown phase and all created resources will be there.
+
+## Lifecycle of tests
+
+Like any other test life cycle, ours are no different and follow the SEVT convention. SEVT stands for Setup, Exercise, Verify and Teardown.
```

Comment:
is there any link to this SEVT convention?

Reply:
http://xunitpatterns.com/Four%20Phase%20Test.html Yes I will also put it there.